tests/integration/ovs_test.py

```diff
@@ -18,103 +18,92 @@
+    bridge = create_bridge_subtree_state(
+        {
+            OVSBridge.FAIL_MODE: '',
+            OVSBridge.MCAST_SNOOPING_ENABLED: False,
+            OVSBridge.RSTP: False,
+            OVSBridge.STP: True,
+        }
+    )
 
-    setup_remove_ovs_bridge_state = {
-        INTERFACES: [
-            {'name': 'ovs-br0', 'type': 'ovs-bridge', 'state': 'absent'}
-        ]
-    }
-    libnmstate.apply(setup_remove_ovs_bridge_state)
-    state = statelib.show_only((desired_state[INTERFACES][0]['name'],))
-    assert not state[INTERFACES]
+    with ovs_bridge(BRIDGE1, bridge) as bridge_state:
+        assertlib.assert_state_match(bridge_state)
+
+    assertlib.assert_absent(BRIDGE1)
+
+
+@pytest.mark.xfail(
+    raises=NmstateLibnmError, reason='https://bugzilla.redhat.com/1724901'
+)
+def test_create_and_remove_ovs_bridge_with_a_system_port(port0_up):
+    port0_name = port0_up[Interface.KEY][0][Interface.NAME]
+
+    bridge = create_bridge_subtree_state()
+    add_port_to_bridge(bridge, port0_name, OVSBridgePortType.SYSTEM)
+
+    with ovs_bridge(BRIDGE1, bridge) as bridge_state:
+        assertlib.assert_state_match(bridge_state)
+
+    assertlib.assert_absent(BRIDGE1)
 
 
 @pytest.mark.xfail(
     raises=NmstateLibnmError, reason='https://bugzilla.redhat.com/1724901'
 )
-def test_create_and_remove_ovs_bridge_with_an_internal_port():
-    state = yaml.load(OVS_BRIDGE_YAML_BASE, Loader=yaml.SafeLoader)
-    state[INTERFACES][0]['bridge']['port'] = [
-        {'name': 'ovs0', 'type': 'internal'}
-    ]
-    ovs_internal_interface_state = {
-        'name': 'ovs0',
-        'type': 'ovs-interface',
-        'state': 'up',
-        'mtu': 1500,
-        'ipv4': {
+def test_create_and_remove_ovs_bridge_with_internal_port_and_static_ip():
+    bridge = create_bridge_subtree_state()
+    add_port_to_bridge(bridge, PORT1, OVSBridgePortType.INTERNAL)
+
+    ipv4_state = {
+        Interface.IPV4: {
             InterfaceIPv4.ENABLED: True,
             InterfaceIPv4.ADDRESS: [
                 {
                     InterfaceIPv4.ADDRESS_IP: '192.0.2.1',
                     InterfaceIPv4.ADDRESS_PREFIX_LENGTH: 24,
                 }
             ],
-        },
-    }
-    state[INTERFACES].append(ovs_internal_interface_state)
-    libnmstate.apply(state, verify_change=False)
-
-    setup_remove_ovs_bridge_state_and_port = {
-        INTERFACES: [
-            {'name': 'ovs-br0', 'type': 'ovs-bridge', 'state': 'absent'},
-            {'name': 'ovs', 'type': 'ovs-interface', 'state': 'absent'},
-        ]
+        }
     }
-    libnmstate.apply(setup_remove_ovs_bridge_state_and_port)
-    state = statelib.show_only(
-        (state[INTERFACES][0]['name'], state[INTERFACES][1]['name'])
-    )
-    assert not state[INTERFACES]
+
+    with ovs_interface(PORT1, ipv4_state, create=False) as iface_state:
+        with ovs_bridge(
+            BRIDGE1, bridge, additional_ifaces=iface_state[Interface.KEY]
```

Comment:
`additional_ifaces` is very ugly. This context does more than it is supposed to (creating an ovs bridge).
It also confuses you and the reader... the returning state includes both the ovs bridge and these additional ifaces. On the following asserts you check it twice.
There is also a problem with the teardown which happens in two transactions (which is a waste).

I guess, for OVS, it fits better to create an internal interface as part of the bridge. Creating an internal interface can be done as stand-alone only by passing a bridge to it, and that will touch both the bridge representation and the internal interface.

I do not like it at all but I do not want to block it. I can refactor it later myself.

Reply:
Done